### DISCUSSION TITLE
sqlite3 dialecte handling savepoints

### DISCUSSION THREAD

**Discussion by JulienPalard** [sqlalchemy/sqlalchemy#7723] [Usage Questions] [answered]
### Describe the use case

For unit test I was in need of an sqlite3 dialect handling nested transactions « from the outside », I mean the tests runner needed to open a transaction and rollback it at the end of the test, while during the test I would let sqlalchemy do its normal transaction stuff.

It does not work with the current implementation because a COMMIT in sqlite [releases all savepoints](https://sqlite.org/lang_savepoint.html), so placing a savepoint before my tests and letting sqlalchemy commiting during the test would release my savepoint ☹

So I was in need for an sqlite3 dialect using SAVEPOINTS only.

### Databases / Backends / Drivers targeted

sqlite3 only.

### Example Use

It would allow nested sessions with arbitrary commits and rollback like:

```python
    with sqlite_with_savepoints_engine.connect():
        with session() as session1:
            assert session1.query(User).count() == 0
            session1.add(User(name="Ada"))
            assert session1.query(User).count() == 1
            with session() as session2:
                assert session2.query(User).count() == 1
                session2.add(User(name="Alan"))
                assert session2.query(User).count() == 2
                with session() as session3, session3.begin():
                    assert session3.query(User).count() == 2
                    session3.add(User(name="Margaret"))
                    assert session3.query(User).count() == 3
                assert session2.query(User).count() == 3
            assert session1.query(User).count() == 1
        with session() as session:
            assert session.query(User).count() == 0
```

(this test case currently pass with my implementation)

### Additional context

I implemented a simple sqlalchemy dialect replacing begin/commit/rollback with savepoints exclusively, it's not even complicated:

```python
from sqlalchemy.dialects.sqlite.pysqlite import SQLiteDialect_pysqlite

class SQLiteWithSavepoints(SQLiteDialect_pysqlite):
    supports_statement_cache = True

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.depth = 0

    def do_begin(self, dbapi_connection):
        self.depth += 1
        dbapi_connection.execute(f"SAVEPOINT s{self.depth}")

    def do_rollback(self, dbapi_connection):
        if self.depth > 0:
            dbapi_connection.execute(f"ROLLBACK TO SAVEPOINT s{self.depth}")
            self.depth -= 1
        else:
            log.debug("Trying to rollback while there's no checkpoint...")

    def do_commit(self, dbapi_connection):
        if self.depth > 0:
            log.debug("RELEASE SAVEPOINT s%d", self.depth)
            dbapi_connection.execute(f"RELEASE SAVEPOINT s{self.depth}")
            self.depth -= 1
        else:
            log.debug("Trying to release checkpoint while there's no checkpoint...")
```

and I feel it would be nice to merge it to the current sqlite3 dialect, but I'm very **very** new to sqlalchemy so I may have missed something... ...how wrong am I?

**Comment by zzzeek** [comment]
hi there

yes, you missed something :)    Python's SQLite dialect has a broken begin()  method, and SQLAlchemy does not try to fix it.  

so to use SAVEPOINTs with SQLite you have to disable this broken method and re-implement BEGIN, but no need to create a new dialect, just use events.   See the section https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#pysqlite-serializable which includes this example:

```py
from sqlalchemy import create_engine, event

engine = create_engine("sqlite:///myfile.db")

@event.listens_for(engine, "connect")
def do_connect(dbapi_connection, connection_record):
    # disable pysqlite's emitting of the BEGIN statement entirely.
    # also stops it from emitting COMMIT before any DDL.
    dbapi_connection.isolation_level = None

@event.listens_for(engine, "begin")
def do_begin(conn):
    # emit our own BEGIN
    conn.exec_driver_sql("BEGIN")
```

**Comment by JulienPalard** [comment]
Hi @zzzeek thanks for looking at my code :)

I don't understand how the implementation https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#pysqlite-serializable from can handle nested transactions, the following:

```python
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker
from sqlalchemy import Column, Text, event, create_engine

engine = create_engine("sqlite://")


@event.listens_for(engine, "connect")
def do_connect(dbapi_connection, connection_record):
    # disable pysqlite's emitting of the BEGIN statement entirely.
    # also stops it from emitting COMMIT before any DDL.
    dbapi_connection.isolation_level = None


@event.listens_for(engine, "begin")
def do_begin(conn):
    # emit our own BEGIN
    conn.exec_driver_sql("BEGIN")


Session = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = declarative_base()


class User(Base):
    __tablename__ = "user"

    name = Column(Text, primary_key=True)


Base.metadata.create_all(bind=engine)

with engine.connect() as connection:
    with Session() as session:
        assert len(session.query(User).all()) == 0
        session.add(User(name="ada"))
        session.flush()
        with Session() as innersession, innersession.begin():
            assert len(innersession.query(User).all()) == 1
```

raises an `sqlite3.OperationalError: cannot start a transaction within a transaction`.

**Comment by zzzeek** [comment]
your example is using a SQLite "memory" database for which SQLAlchemy uses the SingletonThreadPool connection pool.   you can't have two simultaenous transactions present on a SQLite "memory" database without using some more specific APIs.

use a file:

    engine = create_engine("sqlite:///file.db", echo=True)


and for SAVEPOINT use begin_nested()


        with Session() as innersession:
            innersession.begin()
            innersession.begin_nested()
            assert len(innersession.query(User).all()) == 1

**Comment by JulienPalard** [comment]
> and for SAVEPOINT use begin_nested()

That's the whole point: I don't think I can.

My context is writing tests, so from the test code (the outer scope) I want to start a transaction that will rollback everything after the tested code ran. I don't want (and may not even be able to) modify tested code to use `begin_nested` instead of `nested`.

So if I understand correctly, the following does work:
- begin
    - begin_nested

But I need:
- begin_nested
    - begin, or whatever the tested code wants to do, it can even used begin_nested in turn I don't care
    
Currently I probably have `begin`s in the tested code, if it commits (by sending a COMMIT) it will release all savepoints, so I won't be able to rollback anything from the test runner.

> use a file

Same, I don't want to, I'm writing unit tests, there's no point in storing anything on disk.

**Comment by zzzeek** [comment, marked as answer]
To use SAVEPOINT with an ORM test harness, the begin_nested() occurs outside of the scope of where the Session is created.   This is illustrated at  https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites - it's unittest style but the same idea can be adapted to other kinds of fixtures.   This is used by a lot of folks including with SQLite.

**Reply by JulienPalard** [reply]
Oh interesting, I missed this one!

Yet I don't understand:

```
# if the database supports SAVEPOINT (SQLite needs special
# config for this to work), starting a savepoint
# will allow tests to also use rollback within tests
```

which looks wrong, quoting [sqlite3 documentation](https://sqlite.org/lang_savepoint.html):

> The [ROLLBACK](https://sqlite.org/lang_transaction.html) command without a TO clause rolls backs all transactions and leaves the transaction stack empty.

It looks like it has been tested with MySQL but not sqlite3?

If I test it that way (which is very close to the [actual test](https://github.com/sqlalchemy/sqlalchemy/blob/a289e05226e20cc6de9deb479f63297320ef7c5c/test/orm/test_transaction.py#L2391)):

<details>
<summary>(click to expand)</summary>

```python

import logging

logging.basicConfig(level=logging.INFO)
logging.getLogger("sqlalchemy.engine").setLevel(logging.DEBUG)


from contextlib import contextmanager
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker
from sqlalchemy import Column, Text, event, create_engine
from sqlalchemy.pool import StaticPool

engine = create_engine("sqlite://")


Session = sessionmaker(bind=engine)

Base = declarative_base()


class User(Base):
    __tablename__ = "user"

    name = Column(Text, primary_key=True)


Base.metadata.create_all(bind=engine)


@contextmanager
def db_fixture():
    """This simulate what a pytest fixture or what a unittest
    setup/teardown would do.
    """

    connection = engine.connect()
    # begin a non-ORM transaction
    trans = connection.begin()

    # bind an individual Session to the connection
    session = Session(bind=connection)

    nested = connection.begin_nested()

    @event.listens_for(session, "after_transaction_end")
    def end_savepoint(_session, _transaction):
        nonlocal nested
        if not nested.is_active:
            nested = connection.begin_nested()

    yield session

    session.close()

    # rollback - everything that happened with the
    # Session above (including calls to commit())
    # is rolled back.
    trans.rollback()

    # return connection to the Engine
    connection.close()


with db_fixture() as session:
    with session.begin():
        session.add(User(name="Ada"))
        session.flush()

        assert session.query(User).count() == 1
        session.rollback()

    assert session.query(User).count() == 0

    with session.begin():
        session.add(User(name="Ada"))

    assert session.query(User).count() == 1

    with session.begin():
        session.add(User(name="Alan"))
        session.flush()
        assert session.query(User).count() == 2

        session.rollback()
    assert session.query(User).count() == 1
```

</details>

I'm getting a `sqlalchemy.exc.InvalidRequestError: A transaction is already begun on this Session.`.

It really looks like in sqlite3 "commit/rollback" statements [cannot be used inside a checkpoint without leaving the transaction stack empty](https://sqlite.org/lang_savepoint.html), so I really don't see how I can do it without my dialect.

Also my tested code (behind fastapi) creates its own session, so I cannot (easily/cleanly) share the joined session with it, so I don't think it would help if it had worked.

**Reply by zzzeek** [reply]
the SQLite3 documentation is irrelevant here because [this is a bug in the Python pysqlite driver itself, not sqlite](https://bugs.python.org/issue9924).

your sample code is lacking the SQLite event handlers, this is required:

```py
engine = create_engine("sqlite://")

@event.listens_for(engine, "connect")
def do_connect(dbapi_connection, connection_record):
    # disable pysqlite's emitting of the BEGIN statement entirely.
    # also stops it from emitting COMMIT before any DDL.
    dbapi_connection.isolation_level = None

@event.listens_for(engine, "begin")
def do_begin(conn):
    # emit our own BEGIN
    conn.exec_driver_sql("BEGIN")

```

as far as testing on SQLite, you're right, the savepoint-specific recipes are not currently tested on SQLite, and if I apply the recipe to our own tests, we get that same error with our "legacy" tests when using a SQLite memory database, because it's a singleton connection and it's tricky to make sure begin() is only called once with a singleton connection in the pool.    This seems to be an artifact of our internal fixtures though, the recipe itself works fine.   Will fix this today so that we include SQLite savepoint tests with the recipe.

The example that's in the documentation works as expected - here is a complete run that illustrates the SAVEPOINTs working with a SQLite memory database.  start with this script and adapt iteratively:

```py
import unittest
from unittest import TestCase

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import Integer
from sqlalchemy import text
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import sessionmaker

# global application scope.  create Session class, engine
Session = sessionmaker()

engine = create_engine('sqlite://', echo=True)

@event.listens_for(engine, "connect")
def do_connect(dbapi_connection, connection_record):
    # disable pysqlite's emitting of the BEGIN statement entirely.
    # also stops it from emitting COMMIT before any DDL.
    dbapi_connection.isolation_level = None

@event.listens_for(engine, "begin")
def do_begin(conn):
    # emit our own BEGIN
    conn.exec_driver_sql("BEGIN")


Base = declarative_base()

class Foo(Base):
    __tablename__ = "foo"
    id = Column(Integer, primary_key=True)

class Bar(Base):
    __tablename__ = "bar"
    id = Column(Integer, primary_key=True)

Base.metadata.create_all(engine)

class SomeTest(TestCase):
    def setUp(self):
        # connect to the database
        self.connection = engine.connect()

        # begin a non-ORM transaction
        self.trans = self.connection.begin()


        # bind an individual Session to the connection
        self.session = Session(bind=self.connection)


        ###    optional     ###

        # if the database supports SAVEPOINT (SQLite needs special
        # config for this to work), starting a savepoint
        # will allow tests to also use rollback within tests

        self.nested = self.connection.begin_nested()

        @event.listens_for(self.session, "after_transaction_end")
        def end_savepoint(session, transaction):
            if not self.nested.is_active:
                self.nested = self.connection.begin_nested()

        ### ^^^ optional ^^^ ###

    def test_something(self):
        # use the session in tests.

        self.session.add(Foo())
        self.session.commit()

        assert self.session.scalar(text("select count(1) from foo")) == 1

    def test_something_with_rollbacks(self):
        # if the SAVEPOINT steps are taken, then a test can also
        # use session.rollback() and continue working with the database

        self.session.add(Foo())
        self.session.commit()

        assert self.session.scalar(text("select count(1) from foo")) == 1

        self.session.add(Bar())
        self.session.flush()

        assert self.session.scalar(text("select count(1) from foo")) == 1
        assert self.session.scalar(text("select count(1) from bar")) == 1

        self.session.rollback()

        # foo stays committed, bar is rolled back. savepoint worked
        assert self.session.scalar(text("select count(1) from foo")) == 1
        assert self.session.scalar(text("select count(1) from bar")) == 0

        self.session.add(Foo())
        self.session.commit()

        assert self.session.scalar(text("select count(1) from foo")) == 2
        assert self.session.scalar(text("select count(1) from bar")) == 0

    def tearDown(self):
        self.session.close()

        # rollback - everything that happened with the
        # Session above (including calls to commit())
        # is rolled back.
        self.trans.rollback()

        # these will fail if the event handlers are commented out
        assert self.connection.exec_driver_sql("select count(*) from foo").scalar() == 0
        assert self.connection.exec_driver_sql("select count(*) from bar").scalar() == 0

        # return connection to the Engine
        self.connection.close()

if __name__ == "__main__":
    unittest.main()
```

**Reply by zzzeek** [reply]
I've added a patch for enabling the test of this recipe for SQLite with the workaround as given, and there was no issue, I just had to create the engine correctly up front.

**Comment by JulienPalard** [comment]
After playing a game of "7 differences" between my code and yours, it starts to work! The only thing failing in my last tests was:

```python
    with session.begin():
        session.add(User(name="Ada"))
        session.rollback()
    assert session.query(User).count() == 0  # here
    with session.begin():
        session.add(User(name="Ada"))
    assert session.query(User).count() == 1
```

the line marked `here` was starting a new transaction (the `autobegin` thing?), so the following `with session.begin()` was failing (whining about a transation being already started). Now that I understand it, everything starts to work properly \o/

Wow it was not an easy journey learning sqlite transaction internals, pysqlite bugs, sqlalchemy internals all at the same time.

And funny me though I would be able to enhance sqlalchemy on my first day using it, ha, ha.

(Now that I understand better, my dialect was probably very broken, at least in threade contexts, as it was ignoring the session pattern, probably mixing all sessions together very badly, by pushing the transaction lifetime down to the dialect level).

Thanks a lot @zzzeek for your time understanding and replying my questions, I really appreciated it.

**Reply by zzzeek** [reply]
glad you got it working.   yeah, the transaction thing is just not that easy, that's why ive tried to remove different "autocommit" related-options in 2.0 so at least everyone is working with the exact same patterns.  one system that works very consistently is better than 3-6 variants that all behave subtly differently
